File: old/aerox.Runtime.Widgets/Widget.cs
using System.Runtime.InteropServices;
using aerox.Runtime.Math;
using aerox.Runtime.Widgets.Events;

namespace aerox.Runtime.Widgets;

// [StructLayout(LayoutKind.Sequential)]
// public struct WidgetPushConstants
// {
//     public Vector4<float> clip;
//
//     public Vector4<float> extent;
// }
[StructLayout(LayoutKind.Sequential)]
public struct WidgetPushConstants
{
    public Matrix3 Transform;

    public Vector2<float> Size;
}

public abstract class Widget : Disposable
{
    private Size2d? _cachedDesiredSize;

    private WidgetSurface? _cursorUpRoot;
    private Vector2<float> _relativeOffset = 0.0f;
    private Size2d _relativeSize = new();
    private SWidgetsModule _subsystem = SRuntime.Get().GetModule<SWidgetsModule>();

    /// <summary>
    ///     The local angle to render this widget at
    /// </summary>
    public float Angle = 0.0f;

    public WidgetClippingMode ClippingMode = WidgetClippingMode.None;

    /// <summary>
    ///     The pivot used to render this widget. Affects <see cref="Angle" /> and <see cref="Scale" />.
    /// </summary>
    public Vector2<float> Pivot = 0.0f;

    /// <summary>
    ///     The local scale to apply to this widget
    /// </summary>
    public Vector2<float> Scale = 1.0f;

    public bool Hovered { get; private set; }
    public WidgetSurface? Surface { get; private set; }

    public ContainerBase? Parent { get; private set; }

    public WidgetVisibility Visibility { get; set; } = WidgetVisibility.Visible;

    private WidgetPadding _padding = new WidgetPadding();
    /// <summary>
    ///     The Padding For This Widget (Left, Top, Right, Bottom)
    /// </summary>
    public WidgetPadding Padding
    {
        get => _padding;
        set
        {
            _padding = value;
            CheckSize();
        }
    }


    protected bool IsPendingMouseUp => _cursorUpRoot is { Disposed: false };
    
    /// <summary>
    /// Check if this widget is focused by its current surface
    /// </summary>
    public bool Focused => Surface?.FocusedWidget == this;


    /// <summary>
    ///     Computes the relative/local transformation matrix for this widget
    /// </summary>
    /// <returns></returns>
    public Matrix3 ComputeRelativeTransform()
    {
        var m = Matrix3.Identity;
        var offset = _relativeOffset + new Vector2<float>(Padding.Left,Padding.Top);
        
        m = m.Translate(offset + _relativeSize * Pivot);
        m = m.RotateDeg(Angle);
        m = m.Scale(Scale);
        m = m.Translate(offset * Pivot * -1.0f);
        return m;
    }


    public Matrix3 ComputeAbsoluteTransform()
    {
        var parentTransform = Parent?.ComputeAbsoluteTransform() ?? Matrix3.Identity;
        return ComputeRelativeTransform() * parentTransform;
    }

    public void SetVisibility(WidgetVisibility visibility)
    {
        Visibility = visibility;
    }

    public Rect GetRect()
    {
        return new Rect(_relativeOffset, _relativeSize);
    }

    public Rect GetRect(Vector2<float> offset)
    {
        return new Rect(offset + _relativeOffset, _relativeSize);
    }

    public void SetParent(ContainerBase? widget)
    {
        Parent = widget;
    }

    public bool IsSelfHitTestable()
    {
        return Visibility is WidgetVisibility.Visible or WidgetVisibility.VisibleNoHitTestSelf;
    }

    public bool IsChildrenHitTestable()
    {
        return Visibility is WidgetVisibility.Visible or WidgetVisibility.VisibleNoHitTestChildren;
    }

    public bool IsHitTestable()
    {
        return IsSelfHitTestable() || IsChildrenHitTestable();
    }


    public virtual void NotifyAddedToSurface(WidgetSurface widgetSurface)
    {
        Surface = widgetSurface;
        OnAddedToSurface(widgetSurface);
    }

    public virtual void NotifyRemovedFromSurface(WidgetSurface widgetSurface)
    {
        if (Surface == widgetSurface) Surface = null;
        OnRemovedFromSurface(widgetSurface);
    }

    protected virtual void OnAddedToSurface(WidgetSurface widgetSurface)
    {
    }

    protected virtual void OnRemovedFromSurface(WidgetSurface widgetSurface)
    {
    }


    public virtual Widget? ReceiveCursorDown(CursorDownEvent e, DrawInfo info)
    {
        if (IsSelfHitTestable())
            if (OnCursorDown(e))
            {
                _cursorUpRoot = Surface;
                BindCursorUp();
                return this;
            }

        return null;
    }

    protected void BindCursorUp()
    {
        var root = _cursorUpRoot;
        if (root?.Disposed == false) root.OnCursorUp += ReceiveCursorUp;
    }

    protected void UnBindCursorUp()
    {
        var root = _cursorUpRoot;
        if (root?.Disposed == false) root.OnCursorUp -= ReceiveCursorUp;

        _cursorUpRoot = null;
    }

    public virtual void ReceiveCursorUp(CursorUpEvent e)
    {
        UnBindCursorUp();
        OnCursorUp(e);
    }

    public virtual void ReceiveCursorEnter(CursorMoveEvent e, DrawInfo info, List<Widget> items)
    {
        if (!IsSelfHitTestable()) return;
        items.Add(this);
        if (Hovered) return;

        Hovered = true;
        OnCursorEnter(e);
    }

    public virtual bool ReceiveCursorMove(CursorMoveEvent e, DrawInfo info)
    {
        if (IsSelfHitTestable() && OnCursorMove(e)) return true;

        return false;
    }

    public virtual bool ReceiveScroll(ScrollEvent e, DrawInfo info)
    {
        return IsSelfHitTestable() && OnScroll(e);
    }

    protected virtual bool OnCursorDown(CursorDownEvent e)
    {
        return false;
    }

    public virtual void OnCursorUp(CursorUpEvent e)
    {
    }

    protected virtual bool OnCursorMove(CursorMoveEvent e)
    {
        return false;
    }

    protected virtual void OnCursorEnter(CursorMoveEvent e)
    {
    }

    public virtual void ReceiveCursorLeave(CursorMoveEvent e, DrawInfo info)
    {
        if (!Hovered) return;

        Hovered = false;
        OnCursorLeave(e);
    }

    protected virtual void OnCursorLeave(CursorMoveEvent e)
    {
    }

    protected virtual bool OnScroll(ScrollEvent e)
    {
        return false;
    }
    
    public virtual void OnFocus()
    {
    }
    
    public virtual void OnFocusLost()
    {
    }

    protected override void OnDispose(bool isManual)
    {
        UnBindCursorUp();
    }

    public Vector2<float> GetOffset()
    {
        return _relativeOffset;
    }

    public void SetRelativeOffset(Vector2<float> offset)
    {
        _relativeOffset = offset;
    }

    public Size2d GetDrawSize()
    {
        return _relativeSize - new Vector2<float>(Padding.Left + Padding.Right, Padding.Top + Padding.Bottom);
    }

    public virtual void SetDrawSize(Size2d size)
    {
        _relativeSize = size;
    }

    public Size2d GetDesiredSize()
    {
        return _cachedDesiredSize ?? ComputeFinalDesiredSize();
    }

    protected abstract Size2d ComputeDesiredSize();

    private Size2d ComputeFinalDesiredSize() =>
        ComputeDesiredSize() + new Vector2<float>(Padding.Left + Padding.Right, Padding.Top + Padding.Bottom);

    protected virtual bool CheckSize()
    {
        var newSize = ComputeFinalDesiredSize();

        if (newSize.Equals(_cachedDesiredSize)) return false;

        _cachedDesiredSize = newSize;
        Parent?.OnChildResized(this);

        return true;
    }


    public abstract void Collect(WidgetFrame frame, DrawInfo info);
}

File: modules/widgets/lib/aerox/widgets/Widget.cpp
#include "aerox/widgets/Widget.hpp"
#include "aerox/widgets/Container.hpp"
#include "aerox/widgets/Surface.hpp"
#include "aerox/widgets/event/CursorDownEvent.hpp"

namespace aerox::widgets
{
    void Widget::OnAddedToSurface(const Shared<Surface>& widgetSurface)
    {
        _surface = widgetSurface;
    }

    void Widget::OnRemovedFromSurface(const Shared<Surface>& widgetSurface)
    {
        _surface = {};
    }

    bool Widget::OnCursorDown(const Shared<CursorDownEvent>& event)
    {
        return false;
    }

    void Widget::OnCursorUp(const Shared<CursorUpEvent>& event)
    {
    }

    bool Widget::OnCursorMove(const Shared<CursorMoveEvent>& event)
    {
        return false;
    }

    void Widget::OnCursorEnter(const Shared<CursorMoveEvent>& event)
    {
    }

    void Widget::OnCursorLeave(const Shared<CursorMoveEvent>& event)
    {
    }

    bool Widget::OnScroll(const Shared<ScrollEvent>& event)
    {
        return false;
    }

    void Widget::OnFocus()
    {
    }

    void Widget::OnFocusLost()
    {
    }

    Vec2<float> Widget::ComputeFinalDesiredSize()
    {
        return ComputeDesiredSize() + Vec2<float>{padding.left + padding.right, +padding.top + padding.bottom};
    }

    Shared<Surface> Widget::GetSurface() const
    {
        if(auto surface = _surface.lock())
        {
            return surface;
        }

        return  {};
    }

    Shared<Container> Widget::GetParent() const
    {
        if(auto parent = _parent.lock())
        {
            return parent;
        }

        return  {}; 
    }

    Matrix3<float> Widget::ComputeRelativeTransform() const
    {
        auto mat = Matrix3<float>(1.0f);
        auto origin = (_relativeOffset + Vec2<float>(padding.left,padding.top));
        // auto offset = 
        // auto pivotDelta = ;
        // auto totalDelta = offset + _relativeOffset + pivotDelta;
        mat = mat.Translate(GetDrawSize() * pivot * -1.0f).RotateDeg(angle).Translate(origin).Scale(scale);
        // mat = mat.Translate(totalDelta);
        // mat = mat.RotateDeg(angle);
        // mat = mat.Scale(scale);
        // mat = mat.Translate(totalDelta * -1.0f);
        return mat;
    }

    Matrix3<float> Widget::ComputeAbsoluteTransform() const
    {
        if(auto parent = GetParent())
        {
            return ComputeRelativeTransform() * parent->ComputeAbsoluteTransform();
        }
        return ComputeRelativeTransform();
    }

    void Widget::SetVisibility(EVisibility visibility)
    {
        _visibility = visibility;
    }

    EVisibility Widget::GetVisibility() const
    {
        return _visibility;
    }

    Rect Widget::GetRect() const
    {
        return Rect{_relativeOffset,_drawSize};
    }
    

    void Widget::SetParent(const Shared<Container>& container)
    {
        _parent = container;
    }

    void Widget::SetSurface(const Shared<Surface>& surface)
    {
        _surface = surface;
    }

    bool Widget::IsHovered() const
    {
        return _hovered;
    }

    bool Widget::IsHitTestable() const
    {
        return IsSelfHitTestable() || AreChildrenHitTestable();
    }

    bool Widget::IsSelfHitTestable() const
    {
        return _visibility == EVisibility::Visible || _visibility == EVisibility::VisibleNoHitTestChildren;
    }

    bool Widget::AreChildrenHitTestable() const
    {
        return _visibility == EVisibility::Visible || _visibility == EVisibility::VisibleNoHitTestSelf;
    }

    void Widget::BindCursorUp(const Shared<Surface>& surface)
    {
        _cursorUpBinding.UnBind();
        _cursorUpBinding = surface->onCursorUp->Add(this->GetSharedDynamic<Widget>(),&Widget::OnCursorUp);
    }

    void Widget::UnBindCursorUp()
    {
        _cursorUpBinding.UnBind();
    }

    void Widget::NotifyAddedToSurface(const Shared<Surface>& widgetSurface)
    {
        _surface = widgetSurface;
        OnAddedToSurface(widgetSurface);
    }

    void Widget::NotifyRemovedFromSurface(const Shared<Surface>& widgetSurface)
    {
        _surface = {};
        OnRemovedFromSurface(widgetSurface);
    }

    Shared<Widget> Widget::NotifyCursorDown(const Shared<CursorDownEvent>& event, const TransformInfo& transform)
    {
        if (IsSelfHitTestable() && OnCursorDown(event))
        {
            BindCursorUp(event->surface);
            return this->GetSharedDynamic<Widget>();
        }
        return {};
    }

    void Widget::NotifyCursorUp(const Shared<CursorUpEvent>& event)
    {
        UnBindCursorUp();
        OnCursorUp(event);
    }

    void Widget::NotifyCursorEnter(const Shared<CursorMoveEvent>& event, const TransformInfo& transform,
                                   std::vector<Shared<Widget>>& items)
    {
        if(!IsSelfHitTestable()) return;
        items.push_back(this->GetSharedDynamic<Widget>());
        if(IsHovered()) return;

        _hovered = true;
        OnCursorEnter(event);
    }

    bool Widget::NotifyCursorMove(const Shared<CursorMoveEvent>& event, const TransformInfo& transform)
    {
        if (IsSelfHitTestable() && OnCursorMove(event)) return true;

        return false;
    }

    bool Widget::NotifyScroll(const Shared<ScrollEvent>& event, const TransformInfo& transform)
    {
        return IsSelfHitTestable() && OnScroll(event);
    }

    void Widget::NotifyCursorLeave(const Shared<CursorMoveEvent>& event, const TransformInfo& transform)
    {
        if (!IsHovered()) return;

        _hovered = false;
        OnCursorLeave(event);
    }

    Vec2<float> Widget::GetRelativeOffset() const
    {
        return _relativeOffset;
    }

    void Widget::SetRelativeOffset(const Vec2<float>& offset)
    {
        _relativeOffset = offset;
    }

    Vec2<float> Widget::GetDrawSize() const
    {
        return _drawSize;
    }

    void Widget::SetDrawSize(const Vec2<float>& size)
    {
        _drawSize = size;
    }

    Vec2<float> Widget::GetDesiredSize()
    {
        return _cachedDesiredSize.has_value() ? _cachedDesiredSize.value() : ComputeFinalDesiredSize();
    }

    bool Widget::CheckSize()
    {
        auto newSize = ComputeFinalDesiredSize();
        if(_cachedDesiredSize.has_value() && _cachedDesiredSize->NearlyEquals(newSize,0.01f)) return false;

        _cachedDesiredSize = newSize;
        if(auto parent = GetParent())
        {
            parent->OnChildResized(this);
        }

        return true;
    }
}
